### PR TITLE
Added different cross immunity functions to SequencePhenotype

### LINKED ISSUE #3: Implement a new function of Cross-Immunity
### Description 
We want to implement a new function of Cross-Immunity.

### Possible Solution
Use the cross-immunity model from [L&L paper](https://paperpile.com/shared/O4FED8)
image

The main difference between L&L and jhudd's paper is
- jhudd's paper has a max on x_j (the max is across all time points)
- jhudd's paper has a specific form of the cross-immunity distance (exp(- ...))

This difference exists because each strain only appears in one timepoint in the L&L method and its frequency is calculated in a single time bin.

### PR DESCRIPTION
**Pull request by thienktran**
## Description

Updated `riskOfInfection()` in `SequencePhenotype.java`. There are 3 options for `crossImmunityFunction` 
- `"linear"`
  - standard linear mapping function
- `"exponential"`
  - from Erick's [comment](https://github.com/matsengrp/antigen/issues/3#issuecomment-1032911402) on #3 
- `"exponentialSimplified"`
  - simplified equation Erick wrote

The function used to calculate the return value of `riskOfInfection()` can be changed by updating `crossImmunityFunction`  in `parameters.yml`.

Closes #3

## Tests

Added `testRiskOfInfection()` to `TestSequencePhenotype.java`, which passes for all the options of `crossImmunityFunction`. 

Edge cases we tested:

> We compared sequences that 
> - resulted in`testRiskOfInfection()` returning 0
> - shared the same base in half the indices 
>   - `hamming distance == 0.5 * this.sequence.length()`) 
> - were completely different 
>   - `hamming distance == this.sequence.length()`

The main use was tested as well. 

## Checklist:

* [x] The code uses informative and accurate variable and function names
* [x] The functionality is factored out into functions and methods with logical interfaces
* [x] Comments are up to date, document intent, and there are no commented-out code blocks
* [x] Commenting and/or documentation is sufficient for others to be able to understand intent and implementation
* [x] TODOs have been eliminated from the code
* [x] The corresponding issue number (e.g. `#278`) has been searched for in the code to find relevant notes
* [ ] Documentation has been redeployed

(Sorry about all the whitespace. I'm trying to work it out, but IntelliJ isn't cooperating right now)

### CODE REVIEW

File: SequencePhenotype.java [hunk 1]
```diff
@@ -117,14 +119,35 @@ public double distance(Phenotype p) {
...
+
+            switch (Parameters.crossImmunityFunction) {
+                case "linear":
+                    localImmuneRisk = output;
+                    break;
+                case "exponential":
+                    localImmuneRisk =  (1 - exp(-output));
+                    break;
+                case "exponentialSimplified":
+                    localImmuneRisk = 1 - exp(-inputDistance / Parameters.crossImmunityStrength);
```
Comment (matsen) [inline SequencePhenotype.java:143]: Did I have cross-immunity strength here and as an exponent below? Seems a little strange to do that.

**Review by matsen — COMMENTED** [review]
One little comment here, but mostly I don't want to pause at this stage-- let's just get these functions in and then start playing around with them along the lines that Trevor suggested. 🚀

**Comment by huddlej** [comment]
Regarding whitespace, you can hide these changes in the GitHub diffs now:

image

File: SequencePhenotype.java (hunk 1 above)
Comment (thienktran) [reply SequencePhenotype.java:143]: Oh, the exponent below was meant for the first two cases. Thanks for catching that!

(I tested the class with `Parameters.crossImmunityStrength: 1.0`, so it shouldn't have affected the shape of the tree.)

**Review by zorian15 — APPROVED** [review]
looks good to me -- 👍 to merge
